File: typescript/src/booking/application/usecases/bookingRejecting.ts
// Caso d'uso per il rifiuto di una prenotazione.

import type { UseCase } from '@/common/application/usecase.js';
import type { EventPublisher } from '@/common/application/events/eventPublisher.js';
import { Booking } from '../../domain/booking/booking.js';
import { BookingId } from '../../domain/booking/bookingId.js';
import type { BookingEvent } from '../../domain/events/bookingEvent.js';
import type { BookingRejected } from '../../domain/events/bookingResultEvents.js';
import type { BookingRepository } from '../../domain/ports/bookingRepository.js';
import type { RejectBooking } from '../commands/bookingConfirmationCommands.js';
import { requireArgument } from '@/common/utils/requireArgument.js';
import { requireDependency } from '@/common/utils/requireDependency.js';

export class BookingRejecting implements UseCase<RejectBooking, BookingRejected> {
  private readonly _bookingRepository: BookingRepository;
  private readonly _eventPublisher: EventPublisher<BookingEvent>;

  constructor(bookingRepository: BookingRepository, eventPublisher: EventPublisher<BookingEvent>) {
    requireDependency(bookingRepository, "bookingRepository");
    requireDependency(eventPublisher, "eventPublisher");
    this._bookingRepository = bookingRepository;
    this._eventPublisher = eventPublisher;
  }

  invoke(command: RejectBooking): BookingRejected {
    requireArgument(command, 'command');

    const booking = this.findBooking(command.aggregateId);

    const rejected = booking.reject(command.giftCardId, command.amount);

    this._bookingRepository.save(booking);
    this._eventPublisher.publish(rejected);
    return rejected;
  }

  private findBooking(id: BookingId): Booking {
    const booking = this._bookingRepository.findById(id);
    if (booking === null) {
      throw new Error(`Booking not found for rejection: ${id.value}`);
    }
    return booking;
  }
}

File: java/src/main/java/it/giannibombelli/wsc2026/booking/application/usecases/BookingRejecting.java
package it.giannibombelli.wsc2026.booking.application.usecases;

import it.giannibombelli.wsc2026.common.utils.Require;

import it.giannibombelli.wsc2026.booking.domain.booking.Booking;
import it.giannibombelli.wsc2026.booking.domain.events.BookingEvent;
import it.giannibombelli.wsc2026.booking.domain.events.BookingResultEvents;
import it.giannibombelli.wsc2026.booking.domain.ports.BookingRepository;
import it.giannibombelli.wsc2026.common.application.UseCase;
import it.giannibombelli.wsc2026.common.application.events.EventPublisher;

import static it.giannibombelli.wsc2026.booking.application.commands.BookingConfirmationCommands.RejectBooking;
import static java.util.Objects.requireNonNull;

public final class BookingRejecting implements UseCase<RejectBooking, BookingResultEvents.BookingRejected> {
    private final BookingRepository bookingRepository;
    private final EventPublisher<BookingEvent> eventPublisher;

    public BookingRejecting(BookingRepository bookingRepository,
                            EventPublisher<BookingEvent> eventPublisher) {
        this.bookingRepository = requireNonNull(bookingRepository);
        this.eventPublisher = requireNonNull(eventPublisher);
    }

    @Override
    public BookingResultEvents.BookingRejected invoke(RejectBooking cmd) {
        Require.requireArgument(cmd, "command");

        Booking booking = bookingRepository.findById(cmd.aggregateId())
            .orElseThrow(() -> new IllegalStateException("Booking not found for rejection: " + cmd.aggregateId()));

        BookingResultEvents.BookingRejected rejected = booking.reject(cmd.amount());

        bookingRepository.save(booking);
        eventPublisher.publish(rejected);
        return rejected;
    }
}
